src/EnergyPlus/CommandLineInterface.cc

```diff
@@ -816,7 +816,7 @@ namespace CommandLineInterface {
 
             //                                  Heading line found, now looking for Kind
             while (inputFile.good() && !NewHeading) {
-                const auto innerReadResult = inputFile.readLine();
+                const auto &innerReadResult = inputFile.readLine();
```

Comment:
Same here. Code compiles and runs with no diffs but I still question whether correct.

Reply:
I'm inclined to say no ampersand here, but I don't know that it is worth making the change if it is working fine.  I'll try to investigate a bit.

Reply:
I may need to do a wiki page for this case also. The rule should be that the local variable should be of the same type as the function return value. If the function returns a reference, then the local variable that binds its return value should be a reference. If the function returns a value, the local variable should be a value. As for whether the function should be return a value or a (const) reference (again this is only for objects, not for scalars):

* The function can (and probably should) a reference if the object being returned lives beyond the function, i.e., is part of `state`.
* The function should return a value if that value is created within the function, i.e., it does not live beyond function scope.

I looked at the definition of `readLine()` and it looks like it is the latter.

Reply:
Alright, so this could use a quick @amirroth look.  I took the code from around this change and aggressively trim it down to a minimal-ish example.  I have the code on the compiler explorer: https://godbolt.org/z/xfWoY67cc

I am specifically focusing on line 29 of the example, which is:

```
 const auto &innerReadResult = inputFile.readLine();
```

with, or without, the ampersand.

If I **don't** use a reference, here's the code showing up in the compiled assembly for the line:
```
        lea     rax, [rbp-48]
        mov     rdx, QWORD PTR [rbp-56]
        mov     rsi, rdx
        mov     rdi, rax
        call    InputFile::readLine[abi:cxx11]()
```

If I **do** use a reference, here's the code showing up:
```
        lea     rax, [rbp-64]
        mov     rdx, QWORD PTR [rbp-72]
        mov     rsi, rdx
        mov     rdi, rax
        call    InputFile::readLine[abi:cxx11]()
        lea     rax, [rbp-64]
        mov     QWORD PTR [rbp-24], rax
```

Thoughts @amirroth ?

Reply:
I just answered this.

Reply:
For completeness, you also need to look at how `readLine` itself compiles, but the overall point is that this should not be a reference. BTO this issue is orthogonal to `auto`, it's when functions should return pointers or references to things and when they should return values.